pkg/nydussdk/client.go

```diff
@@ -62,191 +149,162 @@ func NewNydusClient(sock string) (Interface, error) {
 }
 
 func (c *NydusdClient) CheckStatus() (*model.DaemonInfo, error) {
-	addr := fmt.Sprintf("http://unix%s", endpointDaemonInfo)
-	resp, err := c.httpClient.Get(addr)
+	url := c.url(endpointDaemonInfo, query{})
+	resp, err := c.httpClient.Get(url)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to do HTTP GET from %s", addr)
+		return nil, errors.Wrapf(err, "get daemon info %s", url)
 	}
 	defer resp.Body.Close()
-	b, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to read status response")
-	}
-	var info model.DaemonInfo
-	if err = json.Unmarshal(b, &info); err != nil {
-		return nil, err
+
+	if succeeded(resp) {
+		var info model.DaemonInfo
+		err = decode(resp, &info)
+		if err != nil {
+			return nil, err
+		}
+		return &info, nil
 	}
-	return &info, nil
+
+	return nil, parseErrorMessage(resp)
 }
 
-func (c *NydusdClient) Umount(sharedMountPoint string) error {
-	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", endpointMount, sharedMountPoint)
-	req, err := http.NewRequest(http.MethodDelete, requestURL, nil)
+func (c *NydusdClient) Umount(mp string) error {
+	query := query{}
+	query.set("mountpoint", mp)
+
+	url := c.url(endpointMount, query)
+	req, err := http.NewRequest(http.MethodDelete, url, nil)
 	if err != nil {
-		return errors.Wrap(err, "failed to create umount request")
+		return errors.Wrap(err, "new http request")
 	}
+
 	resp, err := c.httpClient.Do(req)
 	if err != nil {
-		return errors.Wrapf(err, "failed to do HTTP DELETE to %s", requestURL)
+		return err
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
+
+	if succeeded(resp) {
 		return nil
 	}
-	return handleMountError(resp)
+
+	return parseErrorMessage(resp)
 }
 
 func (c *NydusdClient) GetFsMetric(sharedDaemon bool, sid string) (*model.FsMetric, error) {
-	var getStatURL string
-
+	query := query{}
 	if sharedDaemon {
-		getStatURL = fmt.Sprintf("http://unix%s?id=/%s/fs", endpointMetrics, sid)
-	} else {
-		getStatURL = fmt.Sprintf("http://unix%s", endpointMetrics)
+		query.set("id", "/"+sid)
 	}
 
-	req, err := http.NewRequest(http.MethodGet, getStatURL, nil)
+	url := c.url(endpointMetrics, query)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to create GetFsMetric request")
+		return nil, errors.Wrapf(err, "construct request url %s", url)
 	}
+
 	resp, err := c.httpClient.Do(req)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to do HTTP GET to %s", getStatURL)
+		return nil, err
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
-		return nil, errors.New("got unexpected http status StatusNoContent")
+
+	if !succeeded(resp) {
+		return nil, errors.New("failed to get metrics")
```

Comment:
Why not use `parseErrorMessage(resp)` here?

Reply:
good catch

Reply:
fixed